### PR TITLE
Limit CPU and RAM for Each User Using cgroup

### PR DESCRIPTION
**Pull request by oulfr**
The proposed enhancement involves adding fields to the package modules, allowing administrators to customize CPU, RAM, and SWAP limits for each package in the package module. When a user is created within a specific package, the user's cgroup is updated during both user creation and package updates.

This feature has already been implemented and tested on the latest version of Ubuntu. However, it requires further advanced testing. Importantly, the core of HestiaCP remains unchanged; the modifications are confined to the package and user modules.

If you agree with these changes, I am prepared to submit a pull request to the repository.

### CODE REVIEW

File: bin/v-list-user-package
Comment (jaapmarcus) [inline bin/v-list-user-package:73]: Please keep the spacing in line with the rest

File: bin/v-add-user-package
Comment (jaapmarcus) [inline bin/v-add-user-package:133]: We should include input validation for it

Current state is possible to break things / execute bad things

**Review by jaapmarcus — CHANGES_REQUESTED** [review]
Will set up 

We should also run on rebuild like we do in:
https://github.com/hestiacp/hestiacp/blob/6e0443a61c578bac3e1276b6e6ac977cf8f9dfa7/bin/v-rebuild-user#L52-L55

Update existing packages in:
Existing installs + /install/common/packages/

**Comment by jaapmarcus** [comment]
Haven't tested it yet

**Comment by jaapmarcus** [comment]
Also after enabling and then disabling it keeps the cgroups enabled ... 

Not ideal..

**Comment by oulfr** [comment]
I will try to fix this issues. What do you recommend for enabling and disabling? The user can directly modify the configuration file, and there is currently no way to detect these changes, how it is build for disk quota ?

**Comment by jaapmarcus** [comment]
> I will try to fix this issues. What do you recommend for enabling and disabling? The user can directly modify the configuration file, and there is currently no way to detect these changes, how it is build for disk quota ?

I noticed: 
if [ "$RESOURCES_LIMIT" = 'yes' ]; then
	$BIN/v-update-user-croups "$user"
fi

I assume this should become a setting in Hestia.conf?

If you run v-delete-sys-quota  it will update the config files  and delete the existing  quotas...

v-delete-sys-cgroups should disable them  and

v-add-sys-cgroups can just trigger  a for loop and v-update-user-cgroups user  and loop passed every user and enables them.

Yes they can modify the config files directly but they won't be able to mange the cgroups then anymore ..  And will remain in the limited state... 

Als Edit server need to be updated and RESOURCES_LIMIT need to be added in the sys health config check in:
https://github.com/hestiacp/hestiacp/blob/6e0443a61c578bac3e1276b6e6ac977cf8f9dfa7/func/syshealth.sh#L541-L544

**Comment by oulfr** [comment]
in the v-update-user-package there is a bug:

            for user in $("$BIN/v-list-users" list); do
	            # Fix Bug User Data contain path to the root user
	            USER_DATA="$HESTIA/data/users/$user"
	            check_package=$(grep "PACKAGE='$package'" $USER_DATA/user.conf)
	            if [ -n "$check_package" ]; then
		            "$BIN/v-change-user-package" "$user" "$package" 'yes'
	            fi
            done

There is also some other bug when updating the package the user lose the selected: SH Access.

I will try to fix them

**Comment by oulfr** [comment]
All issue are fixed, this is how the test is done:

1. Install the stress package: sudo apt install -y stress.
2. Create two packages: one limited with 50%, 20 seconds, 1GB memory, and 1GB swap; and another unlimited without any constraints.
3. Create two users: one limited with the restricted package and another unlimited with the unrestricted package.
4. Open three terminals and connect with root, limited, and unlimited users.
5. In the root terminal, run: top
6. In the limited and unlimited user terminals, run: stress -m 1 --vm-bytes 4G --vm-keep.
7. Go to the limited package and change the CPU value or memory in the root terminal; you will see your changes immediately.

**Comment by oulfr** [comment]
The test won't pass because both $HESTIA/data/packages/default.pkg and $HESTIA/data/packages/system.pkg need to be patched with some extra field. what do you suggest for this ?

**Comment by BrunoHsr** [comment]
Hello, how do I update so that my hestiacp has these settings?

**Comment by jaapmarcus** [comment]
@oulfr  Why did your closed / deleted the PR ?

**Comment by jaapmarcus** [comment]
> Hello, how do I update so that my hestiacp has these settings?

Wait until it has been merged it was still in active development...

**Comment by oulfr** [comment]
can you contact me here:aoulmderat@gmail.com

**Comment by jaapmarcus** [comment]
> can you contact me here:[aoulmderat@gmail.com](mailto:aoulmderat@gmail.com)

Send you an email from info@hestiacp.com

**Comment by oulfr** [comment]
I will try to push another PR

**Pull request closed without merging**
